src/trel_dnssd/trel_dnssd.cpp

```diff
@@ -0,0 +1,542 @@
+        [this](const std::string &aType, const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo) {
+            OnTrelServiceInstanceResolved(aType, aInstanceInfo);
+        },
+
+        nullptr);
+
+    if (IsReady())
+    {
+        mPublisher.SubscribeService(kTrelServiceName, "");
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::StopBrowse(void)
+{
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Stop browsing %s service.", kTrelServiceName);
+    assert(mSubscriberId > 0);
+
+    mPublisher.RemoveSubscriptionCallbacks(mSubscriberId);
+    mSubscriberId = 0;
+
+    if (IsReady())
+    {
+        mPublisher.UnsubscribeService(kTrelServiceName, "");
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::RegisterService(uint16_t aPort, const uint8_t *aTxtData, uint8_t aTxtLength)
+{
+    assert(aPort > 0);
+    assert(aTxtData != nullptr);
+
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Register %s service: port=%u, TXT=%d bytes", kTrelServiceName, aPort, aTxtLength);
+    otbrDump(OTBR_LOG_DEBUG, OTBR_LOG_TAG, "TXT", aTxtData, aTxtLength);
+
+    if (mRegisterInfo.IsValid() && IsReady())
+    {
+        UnpublishTrelService();
+    }
+
+    mRegisterInfo.Assign(aPort, aTxtData, aTxtLength);
+
+    if (IsReady())
+    {
+        PublishTrelService();
+    }
+
+exit:
+    return;
+}
+
+void TrelDnssd::UnregisterService(void)
+{
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("Remove %s service.", kTrelServiceName);
+    assert(mRegisterInfo.IsValid());
+
+    if (IsReady())
+    {
+        UnpublishTrelService();
+    }
+
+    mRegisterInfo.Clear();
+
+exit:
+    return;
+}
+
+void TrelDnssd::OnMdnsPublisherReady(void)
+{
+    VerifyOrExit(IsInitialized());
+
+    otbrLogDebug("mDNS Publisher is Ready");
+    mMdnsPublisherReady = true;
+    RemoveAllPeers();
+
+    if (mRegisterInfo.IsPublished())
+    {
+        mRegisterInfo.mInstanceName = "";
+    }
+
+    OnBecomeReady();
+
+exit:
+    return;
+}
+
+void TrelDnssd::OnTrelServiceInstanceResolved(const std::string &                            aType,
+                                              const Mdns::Publisher::DiscoveredInstanceInfo &aInstanceInfo)
+{
+    VerifyOrExit(StringUtils::EqualCaseInsensitive(aType, kTrelServiceName));
+    VerifyOrExit(aInstanceInfo.mNetifIndex == mTrelNetifIndex);
+
+    if (aInstanceInfo.mRemoved)
+    {
+        OnTrelServiceInstanceRemoved(aInstanceInfo.mName);
+    }
+    else
+    {
+        OnTrelServiceInstanceAdded(aInstanceInfo);
+    }
+
+exit:
+    return;
+}
+
+std::string TrelDnssd::GetTrelInstanceName(void)
+{
+    const otExtAddress *extaddr = otLinkGetExtendedAddress(mNcp.GetInstance());
+    std::string         name;
+
+    name.reserve(sizeof(extaddr) * 2);
+
+    for (uint8_t b : extaddr->m8)
+    {
+        char buf[3];
+
+        sprintf(buf, "%02x", b);
+        name.push_back(buf[0]);
+        name.push_back(buf[1]);
+    }
+
+    assert(name.length() == sizeof(extaddr) * 2);
```

Comment:
Changed to use Utils::Bytes2Hex. However, the Utils::Bytes2Hex itself is not implemented as efficient as it can be. 
Will refactor `Utils::Bytes2Hex` in a future PR.